tf/tfprocess.py

```diff
@@ -486,36 +607,130 @@ def construct_net(self, planes):
-        self.weights.append(W_fc2)
-        self.weights.append(b_fc2)
+                                   output_channels=32,
+                                   name="value_head")
+        h_conv_val_flat = tf.reshape(conv_val, [-1, 32 * 8 * 8])
+        W_fc2 = weight_variable("w_fc_2", [32 * 8 * 8, 256])
+        b_fc2 = bias_variable("b_fc_2", [256])
+        self.add_weights(W_fc2)
+        self.add_weights(b_fc2)
         h_fc2 = tf.nn.relu(tf.add(tf.matmul(h_conv_val_flat, W_fc2), b_fc2))
-        W_fc3 = weight_variable([128, 1], name='fc3/weight')
-        b_fc3 = bias_variable([1], name='fc3/bias')
-        self.weights.append(W_fc3)
-        self.weights.append(b_fc3)
-        h_fc3 = tf.nn.tanh(tf.add(tf.matmul(h_fc2, W_fc3), b_fc3), name='value_head')
+        W_fc3 = weight_variable("w_fc_3", [256, 1])
+        b_fc3 = bias_variable("b_fc_3", [1])
+        self.add_weights(W_fc3)
+        self.add_weights(b_fc3)
+        h_fc3 = tf.nn.tanh(tf.add(tf.matmul(h_fc2, W_fc3), b_fc3))
 
         return h_fc1, h_fc3
+
+    def snap_save(self):
+        # Save a snapshot of all the variables in the current graph.
+        if not hasattr(self, 'save_op'):
+            save_ops = []
+            rest_ops = []
+            for var in self.weights:
+                if isinstance(var, str):
+                    var = tf.get_default_graph().get_tensor_by_name(var)
+                name = var.name.split(':')[0]
+                v = tf.Variable(var, name='save/'+name, trainable=False)
+                save_ops.append(tf.assign(v, var))
+                rest_ops.append(tf.assign(var, v))
+            self.save_op = tf.group(*save_ops)
+            self.restore_op = tf.group(*rest_ops)
+        self.session.run(self.save_op)
+
+    def snap_restore(self):
+        # Restore variables in the current graph from the snapshot.
+        self.session.run(self.restore_op)
+
+    def save_swa_network(self, steps, path, leela_path, data):
+        # Sample 1 in self.swa_c of the networks. Compute in this way so
+        # that it's safe to change the value of self.swa_c
+        rem = self.session.run(tf.assign_add(self.swa_skip, -1))
+        if rem > 1:
+            return
+        self.swa_skip.load(self.swa_c, self.session)
+
+        # Add the current weight vars to the running average.
+        num = self.session.run(self.swa_accum_op)
+        num = min(num, self.swa_max_n)
+        self.swa_count.load(float(num), self.session)
+
+        swa_path = path + "-swa-" + str(int(num)) + "-" + str(steps)
+
+        # save the current network.
+        self.snap_save()
+        # Copy the swa weights into the current network.
+        self.session.run(self.swa_load_op)
+        if self.swa_recalc_bn:
+            print("Refining SWA batch normalization")
+            for _ in range(200):
+                batch = next(data)
+                self.session.run(
+                    [self.loss, self.update_ops],
```

Comment:
Either way, as people have already mentioned in the Discord chat, I don't think this loop is necessary or even better than not adjusting BN params. It would be sensible if our architecture had Conv-ReLU-BN because the ReLU would affect the distribution of activations after SWA non-trivially, but our architecture with Conv-BN-ReLU means that BN params can be folded into the Conv layer easily, so I think not fine-tuning BN running mean and variances might actually the right thing to do.